File: p2/p261/ValidTree.java
package leetcode.p2.p261;

import java.util.*;

public class ValidTree {
    public boolean validTree(int n, int[][] edges) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        for (int i = 0; i < n; i++) {
            graph.put(i, new HashSet<>());
        }

        for (int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }

        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();
        queue.offer(0);
        visited.add(0);
        while (!queue.isEmpty()) {
            int node = queue.poll();
            for (int next : graph.get(node)) {
                if (visited.contains(next)) {
                    return false;
                }
                graph.get(next).remove(node);
                queue.offer(next);
                visited.add(next);
            }
        }
        return visited.size() == n;
    }
}

File: p2/p261/valid_tree.py
from typing import List


class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n - 1:
            return False
        if n == 1:
            return True
        graph_map = {}
        for edge in edges:
            if edge[0] not in graph_map:
                graph_map[edge[0]] = set()
            graph_map[edge[0]].add(edge[1])

            if edge[1] not in graph_map:
                graph_map[edge[1]] = set()
            graph_map[edge[1]].add(edge[0])

        visited = set()

        stack = [edges[0][0]]
        while stack:
            node = stack.pop()
            visited.add(node)

            for next_node in graph_map[node]:
                if next_node not in visited:
                    stack.append(next_node)

        if len(visited) != n:
            return False

        return True


if __name__ == '__main__':
    sol = Solution()
    print(sol.validTree(5, [[0, 1], [0, 2], [0, 3], [1, 4]]))
